### ISSUE TITLE
Multiple Calls to MLC.addProcess in for loop -> error

### ISSUE THREAD

**Issue by alexmorley** [flatironinstitute/mountainlab-js#45]
e.g.
```js
async function convert_firings(MLC, dataset_dirname, output_dirname, params, temp_dir) {
  for (tet = 1; tet < params.tetrodes.length+1; tet++) {
    MLC.addProcess({
      processor_name: 'dd.convert_firings',
      inputs: {
        firings: output_dirname + '/firings.mda',
        params: temp_dir + 'params.json'
      },
      outputs: {
        res_fname: output_dirname + '/res.' + tet,
        clu_fname: output_dirname + '/clu.' + tet
      },
      parameters: {
        tetrode:tet
      }
    })
  };
  await MLC.run();
}

```

Errors like
```
[...]
Error in find: Unexpected token { in JSON at position 10746 
[...]
```

I think its a race condition. Maybe two processes trying to write to the database at the same time?

Is there a lock on the database?

**Comment by alexmorley** [comment]
Yeah, you'd have thought that'd be like no. 1 requirement for any database... Don't know what progress you've made yet but the easiest fix I can think of is for a process to check for a db.lock file, if its present then wait until its not, then write a db.lock file, then to the actual db, then remove db.lock.

**Comment by magland** [comment]
Thanks Alex. After exploring various solutions, looking at the source code for diskd (and having some painful experiences)... I decided to roll my own (which does use a locking mechanism similar to what you suggest).

This needs to be tested!

I've just pushed the changes, and you can see the source code here: https://github.com/flatironinstitute/mountainlab-js/blob/master/mlproc/safedatabase.js

**Comment by magland** [comment]
This passed, which is good news: https://travis-ci.org/flatironinstitute/mountainlab-js-tests

**Comment by tjd2002** [comment]
I don’t think it would be *too* painful to go back to Mongo (user-local installs are easy to spin up) if this turns out to be a bigger pain point.
